File: client/AndroidStudio/CloudBlackboard/app/src/main/java/com/beautyli/app/cloudblackboard/MyClient.java
package com.beautyli.app.cloudblackboard;

import android.util.Base64;

import java.io.*;
import java.net.Socket;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;
import java.util.zip.Deflater;
import java.util.zip.Inflater;

import com.google.gson.JsonIOException;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.JsonSyntaxException;

public class MyClient {

    public static String mServer_IP;
    public static int mServer_Port;
    public static long mCurDBVersion = 0;

    public String mLastReason;

    private JsonParser mJsonParser = new JsonParser();
    private Socket mSocket = new Socket();
    private DataInputStream mInput;
    private DataOutputStream mOutput;
    private RC4 mRC4 = new RC4();

    private final String mRC4_BaseKey = "CloudBlackboard_rc4_base_key";

    public synchronized boolean connect() {
        try {
            mSocket = new Socket(mServer_IP, mServer_Port);
            mInput = new DataInputStream(new BufferedInputStream(mSocket.getInputStream()));
            mOutput = new DataOutputStream(mSocket.getOutputStream());
        } catch (IOException e) {
            return false;
        }

        return mSocket.isConnected();
    }

    public synchronized void close() {
        try {
            mSocket.close();
        } catch (IOException ignored) {
        }
    }

    public synchronized boolean logon(String user, String pin) {
        mLastReason = "null";
        try {
            if (!mSocket.isConnected()) {
                return false;
            }

            byte[] rand_buffer = new byte[8];
            if (mInput.read(rand_buffer) < rand_buffer.length) {
                return false;
            }

            //拼接密钥
            byte[] rc4_key = (mRC4_BaseKey + pin).getBytes("UTF-8");

            //加密随机数
            mRC4.setKey(rc4_key);
            byte[] enc_rand = mRC4.doEnc(rand_buffer);
            String challenge = Base64.encodeToString(enc_rand, Base64.DEFAULT).trim();

            JsonObject request = new JsonObject();
            request.addProperty("transaction", "logon");
            request.addProperty("user", user);
            request.addProperty("challenge", challenge);

            String request_string = request.toString();
            byte[] request_data = request_string.getBytes("UTF-8");

            if (!sendPack(request_data)) {
                return false;
            }

            JsonObject respond = doRequest(null);
            if (respond == null) {
                return false;
            }

            if (respond.has("reason")) {
                mLastReason = respond.get("reason").getAsString();
            }
            String result = respond.get("result").getAsString();
            if (0 != result.compareTo("success"))
                return false;

            byte[] session_key = Base64.decode(respond.get("session_key").getAsString(), Base64.DEFAULT);
            mRC4.setKey(session_key);
            return true;
        }
        catch (IOException e) {
            return false;
        }
    }

    public synchronized boolean register(String user, String pin) {
        mLastReason = "null";
        try {
            if (!mSocket.isConnected()) {
                return false;
            }

            byte[] rand_buffer = new byte[8];
            if (mInput.read(rand_buffer) < rand_buffer.length) {
                return false;
            }

            //密钥
            byte[] rc4_key = mRC4_BaseKey.getBytes("UTF-8");

            //加密随机数
            mRC4.setKey(rc4_key);
            byte[] enc_rand = mRC4.doEnc(rand_buffer);
            String challenge = Base64.encodeToString(enc_rand, Base64.DEFAULT);

            //加密PIN
            String enc_pin = Base64.encodeToString(mRC4.doEnc(pin.getBytes("UTF-8")), Base64.DEFAULT);

            JsonObject request = new JsonObject();
            request.addProperty("transaction", "register");
            request.addProperty("user", user);
            request.addProperty("challenge", challenge);
            request.addProperty("pin", enc_pin);

            String request_string = request.toString();
            byte[] request_data = request_string.getBytes("UTF-8");
            if (!sendPack(request_data)) {
                return false;
            }

            JsonObject respond = doRequest(null);
            if (respond == null) {
                return false;
            }

            if (respond.has("reason")) {
                mLastReason = respond.get("reason").getAsString();
            }
            String result = respond.get("result").getAsString();
            return result.compareTo("success") == 0;
        }
        catch (IOException e) {
            return false;
        }
    }

    public synchronized JsonObject getAllNotes() {
        mLastReason = "null";
        if (!mSocket.isConnected()) {
            return null;
        }

        JsonObject request = new JsonObject();
        request.addProperty("transaction", "get_new_notes");
        request.addProperty("sheet", "default");
        request.addProperty("client_db_version", mCurDBVersion);

        JsonObject respond = doRequest(request);
        if (null == respond) {
            return null;
        }

        if (respond.has("reason")) {
            mLastReason = respond.get("reason").getAsString();
        }
        String result = respond.get("result").getAsString();
        if (0 == result.compareTo("success")) {
            mCurDBVersion = respond.get("ver").getAsLong();
            return respond;
        } else {
            return null;
        }
    }

    public synchronized JsonObject uploadNotes(String noteID, String noteText) {
        mLastReason = "null";
        if (!mSocket.isConnected()) {
            return null;
        }

        JsonObject request = new JsonObject();
        request.addProperty("transaction", "put_note");
        request.addProperty("sheet", "default");
        request.addProperty("note_id", noteID);
        request.addProperty("time", System.currentTimeMillis() / 1000);
        request.addProperty("text", noteText);

        JsonObject respond = doRequest(request);
        if (null == respond) {
            return null;
        }

        if (respond.has("reason")) {
            mLastReason = respond.get("reason").getAsString();
        }
        String result = respond.get("result").getAsString();
        if (0 == result.compareTo("success")) {
            return respond;
        } else {
            return null;
        }
    }

    public synchronized JsonObject deleteNotes(String noteID) {
        mLastReason = "null";
        if (!mSocket.isConnected()) {
            return null;
        }

        JsonObject request = new JsonObject();
        request.addProperty("transaction", "delete_note");
        request.addProperty("sheet", "default");
        request.addProperty("note_id", noteID);

        JsonObject respond = doRequest(request);
        if (null == respond) {
            return null;
        }

        if (respond.has("reason")) {
            mLastReason = respond.get("reason").getAsString();
        }
        String result = respond.get("result").getAsString();
        if (0 == result.compareTo("success")) {
            return respond;
        } else {
            return null;
        }
    }

    private synchronized JsonObject doRequest(JsonObject request) {
        try {
            if (!mSocket.isConnected()) {
                return null;
            }

            if (request != null) {
                String request_string = request.toString();
                byte[] request_data = request_string.getBytes("UTF-8");
                byte[] send_data = packSendData(request_data);
                sendPack(send_data);
            }

            byte[] receive_data = receivePack();
            if (null == receive_data) {
                return null;
            }

            byte[] respond_data = unpackReceiveData(receive_data);
            if(respond_data == null) {
                return null;
            }
            String respond_string = new String(respond_data, "UTF-8");
            return mJsonParser.parse(respond_string).getAsJsonObject();

        } catch (UnsupportedEncodingException | JsonIOException | JsonSyntaxException e) {
            return null;
        }
    }

    private byte[] receivePack() {
        try {
            if (!mSocket.isConnected()) {
                return null;
            }

            byte[] pack_size_buffer = new byte[4];
            if (mInput.read(pack_size_buffer) < pack_size_buffer.length) {
                return null;
            }

            int pack_size = (pack_size_buffer[0] & 0xff) |
                    (pack_size_buffer[1] & 0xff) << 8 |
                    (pack_size_buffer[2] & 0xff) << 16 |
                    (pack_size_buffer[3] & 0xff) << 24;
            if (pack_size < 16 || pack_size > 1024 * 1024 * 32) {
                return null;
            }

            byte[] receive_data = new byte[pack_size];
            int receive_len = 0;
            while (receive_len < pack_size) {
                byte[] receive_buffer = new byte[pack_size - receive_len];
                int temp_len = mInput.read(receive_buffer);
                if (temp_len <= 0) {
                    return null;
                }
                System.arraycopy(receive_buffer, 0, receive_data, receive_len, temp_len);
                receive_len += temp_len;
            }
            return receive_data;

        } catch (IOException e) {
            return null;
        }
    }

    private boolean sendPack(byte[] data) {
        try {
            if (!mSocket.isConnected()) {
                return false;
            }

            int data_len = data.length;
            byte[] b_data_len = new byte[4];
            b_data_len[0] = (byte) (data_len & 0xff);
            b_data_len[1] = (byte) ((data_len >> 8) & 0xff);
            b_data_len[2] = (byte) ((data_len >> 16) & 0xff);
            b_data_len[3] = (byte) ((data_len >> 24) & 0xff);

            mOutput.write(b_data_len);
            mOutput.write(data);
            return true;

        } catch (IOException e) {
            return false;
        }
    }

    private byte[] unpackReceiveData(byte[] data) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            int hash_size = digest.getDigestLength();
            //解密数据
            byte[] dec_data = mRC4.doEnc(data);
            if (dec_data.length > hash_size) {
                byte[] receive_hash = new byte[hash_size];
                byte[] receive_data = new byte[dec_data.length - hash_size];
                System.arraycopy(dec_data, 0, receive_hash, 0, hash_size);
                System.arraycopy(dec_data, hash_size, receive_data, 0, receive_data.length);
                //校验hash
                byte[] hash = digest.digest(receive_data);
                if (Arrays.equals(hash, receive_hash)) {
                    // 解压缩
                    return decompress(receive_data);
                }
            }
        } catch (NoSuchAlgorithmException e) {
            return null;
        }
        return null;
    }

    private byte[] packSendData(byte[] data) {
        try {
            // 压缩
            byte[] zip_data = compress(data);
            if(zip_data == null) {
                return null;
            }
            // Hash
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(zip_data);

            byte[] send_data = new byte[hash.length + zip_data.length];
            System.arraycopy(hash, 0, send_data, 0, hash.length);
            System.arraycopy(zip_data, 0, send_data, hash.length, zip_data.length);

            return mRC4.doEnc(send_data);

        } catch (NoSuchAlgorithmException e) {
            return null;
        }
    }

    private byte[] compress(byte[] data) {
        byte[] output;
        Deflater compresser = new Deflater();
        compresser.reset();
        compresser.setInput(data);
        compresser.finish();
        ByteArrayOutputStream bos = new ByteArrayOutputStream(data.length);
        try {
            byte[] buf = new byte[2048];
            while (!compresser.finished()) {
                int i = compresser.deflate(buf);
                bos.write(buf, 0, i);
            }
            output = bos.toByteArray();
        } catch (Exception e) {
            return null;
        } finally {
            try {
                bos.close();
            } catch (IOException ignored) {
            }
        }
        compresser.end();
        return output;
    }

    private byte[] decompress(byte[] data) {
        byte[] output;
        Inflater decompresser = new Inflater();
        decompresser.reset();
        decompresser.setInput(data);
        ByteArrayOutputStream bos = new ByteArrayOutputStream(data.length);
        try {
            byte[] buf = new byte[2048];
            while (!decompresser.finished()) {
                int i = decompresser.inflate(buf);
                bos.write(buf, 0, i);
            }
            output = bos.toByteArray();
        } catch (Exception e) {
            return null;
        } finally {
            try {
                bos.close();
            } catch (IOException ignored) {
            }
        }
        decompresser.end();
        return output;
    }
}

File: client/VS2017/CloudBlackboard/CloudBlackboard/MyClient.cs
using System;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Security.Cryptography;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using zlib;

namespace CloudBlackboard
{
    public class MyClient
    {
        private RC4 mRC4 = new RC4();

        private const string mRC4_BaseKey = "CloudBlackboard_rc4_base_key";

        public string mServerIP = "beautyli.cn";
        public int mServerPort = 16821;

        static readonly int mConnectTimeout = 10 * 1000;

        Socket mSocket;

        public bool Connect()
        {
            lock (this)
            {
                if (mSocket != null)
                {
                    mSocket.Close();
                }
                mSocket = new Socket(SocketType.Stream, ProtocolType.Tcp);

                try
                {
                    //IPEndPoint end_point = new IPEndPoint(Dns.GetHostEntry(mServerIP).AddressList[0], mServerPort);
                    IPEndPoint end_point = new IPEndPoint(IPAddress.Parse(mServerIP), mServerPort);

                    IAsyncResult connResult = mSocket.BeginConnect(end_point, null, null);
                    connResult.AsyncWaitHandle.WaitOne(mConnectTimeout, true);
                    if (!connResult.IsCompleted)
                    {
                        mSocket.Close();
                        return false;
                    }
                }
                catch (Exception)
                {
                    mSocket.Close();
                    return false;
                }

                if (!mSocket.Connected)
                {
                    mSocket.Close();
                    return false;
                }

                mSocket.ReceiveTimeout = mConnectTimeout / 2;
                return true;
            }
        }

        int Receive(byte[] buf)
        {
            try
            {
                int r = mSocket.Receive(buf);
                if (r <= 0)
                {
                    mSocket.Close();
                }
                return r;
            }
            catch (Exception)
            {
                mSocket.Close();
                return -1;
            }
        }

        public bool Logon(string user, string pin)
        {
            lock (this)
            {
                if (!mSocket.Connected)
                {
                    return false;
                }

                byte[] rand_buffer = new byte[8];
                if (Receive(rand_buffer) < rand_buffer.Length)
                {
                    return false;
                }

                //拼接密钥
                byte[] rc4_key = System.Text.Encoding.UTF8.GetBytes(mRC4_BaseKey + pin);

                //加密随机数
                mRC4.SetKey(rc4_key);
                byte[] enc_rand = mRC4.DoEnc(rand_buffer);
                string challenge = Convert.ToBase64String(enc_rand);

                JObject request = new JObject
                {
                    ["transaction"] = "logon",
                    ["user"] = user,
                    ["challenge"] = challenge
                };

                string request_string = JsonConvert.SerializeObject(request);
                byte[] request_data = System.Text.Encoding.UTF8.GetBytes(request_string);

                if (!SendPack(request_data))
                {
                    return false;
                }

                JObject respond = DoRequest(null);
                if (respond == null)
                {
                    return false;
                }

                string result = (string)respond["result"];
                if (result != "success")
                    return false;

                byte[] session_key = Convert.FromBase64String((string)respond["session_key"]);
                mRC4.SetKey(session_key);
                return true;
            }
        }

        public bool Register(string user, string pin, ref string reason)
        {
            lock (this)
            {
                if (!mSocket.Connected)
                {
                    return false;
                }

                byte[] rand_buffer = new byte[8];
                if (Receive(rand_buffer) < rand_buffer.Length)
                {
                    return false;
                }

                //密钥
                byte[] rc4_key = System.Text.Encoding.UTF8.GetBytes(mRC4_BaseKey);

                //加密随机数
                mRC4.SetKey(rc4_key);
                byte[] enc_rand = mRC4.DoEnc(rand_buffer);
                string challenge = Convert.ToBase64String(enc_rand);

                //加密PIN
                string enc_pin = Convert.ToBase64String(mRC4.DoEnc(System.Text.Encoding.UTF8.GetBytes(pin)));

                JObject request = new JObject
                {
                    ["transaction"] = "register",
                    ["user"] = user,
                    ["pin"] = enc_pin,
                    ["challenge"] = challenge
                };

                string request_string = JsonConvert.SerializeObject(request);
                byte[] request_data = System.Text.Encoding.UTF8.GetBytes(request_string);
                if (!SendPack(request_data))
                {
                    return false;
                }

                JObject respond = DoRequest(null);
                if (respond == null)
                {
                    return false;
                }

                string result = (string)respond["result"];
                if(respond.ContainsKey("reason"))
                {
                    reason = (string)respond["reason"];
                }
                return result == "success";
            }
        }

        private JObject DoRequest(JObject request)
        {
            lock (this)
            {
                try
                {
                    if (!mSocket.Connected)
                    {
                        return null;
                    }

                    if(request != null)
                    {
                        string request_string = JsonConvert.SerializeObject(request);
                        byte[] request_data = System.Text.Encoding.UTF8.GetBytes(request_string);
                        byte[] send_data = PackSendData(request_data);
                        SendPack(send_data);
                    }

                    byte[] receive_data = ReceivePack();
                    if(null == receive_data)
                    {
                        return null;
                    }

                    byte[] respond_data = UnpackReceiveData(receive_data);
                    string respond_string = System.Text.Encoding.UTF8.GetString(respond_data);
                    JObject respond = (JObject)JsonConvert.DeserializeObject(respond_string);
                    return respond;
                }
                catch (Exception)
                {
                    return null;
                }
            }
        }

        public JObject GetNewNotes(ref int clientDBVersion)
        {
            JObject request = new JObject
            {
                ["transaction"] = "get_new_notes",
                ["sheet"] = "default",
                ["client_db_version"] = clientDBVersion,
            };

            JObject respond = DoRequest(request);
            if (null == respond)
            {
                return null;
            }

            string result = (string)respond["result"];
            if (result == "success")
            {
                clientDBVersion = (int)respond["ver"];
                return respond;
            }
            else
            {
                return null;
            }
        }

        public bool PutNote(string note_id, string text, string filePath = null)
        {
            JObject request = new JObject
            {
                ["transaction"] = "put_note",
                ["sheet"] = "default",
                ["note_id"] = note_id,
                ["text"] = text,
                ["time"] = Convert.ToInt32(DateTime.UtcNow.Subtract(DateTime.Parse("1970-1-1")).TotalSeconds)
            };

            JObject respond = DoRequest(request);
            if (null == respond)
            {
                return false;
            }

            string result = (string)respond["result"];
            if (result != "success")
            {
                return false;
            }
            return true;
        }

        public bool DeleteNote(string note_id)
        {
            JObject request = new JObject
            {
                ["transaction"] = "delete_note",
                ["sheet"] = "default",
                ["note_id"] = note_id
            };

            JObject respond = DoRequest(request);
            if (null == respond)
            {
                return false;
            }

            string result = (string)respond["result"];
            if (result != "success")
            {
                string reason = (string)respond["reason"];
                if (reason != "note not existed")
                {
                    return false;
                }
            }
            return true;
        }

        private static bool __MemoryCompare(byte[] b1, byte[] b2, int len)
        {
            if (b1.Length < len || b2.Length < len)
            {
                return false;
            }
            for (int i = 0; i < len; i++)
            {
                if (b1[i] != b2[i])
                {
                    return false;
                }
            }
            return true;
        }

        private byte[] ReceivePack()
        {
            try
            {
                if (!mSocket.Connected)
                {
                    return null;
                }

                byte[] pack_size_buffer = new byte[4];
                if (Receive(pack_size_buffer) < pack_size_buffer.Length)
                {
                    return null;
                }

                int pack_size = pack_size_buffer[0] +
                    pack_size_buffer[1] * 256 +
                    pack_size_buffer[2] * 256 * 256 +
                    pack_size_buffer[3] * 256 * 256 * 256;
                if (pack_size < 16 || pack_size > 1024 * 1024 * 32)
                {
                    return null;
                }

                byte[] receive_data = new byte[pack_size];
                int receive_len = 0;
                while (receive_len < pack_size)
                {
                    byte[] receive_buffer = new byte[pack_size - receive_len];
                    int temp_len = Receive(receive_buffer);
                    if (temp_len <= 0)
                    {
                        return null;
                    }
                    Array.Copy(receive_buffer, 0, receive_data, receive_len, temp_len);
                    receive_len += temp_len;
                }

                return receive_data;
            }
            catch (Exception)
            {
                return null;
            }
        }

        private bool SendPack(byte[] data)
        {
            try
            {
                if (!mSocket.Connected)
                {
                    return false;
                }

                int data_len = data.Length;
                byte[] b_data_len = new byte[4];
                b_data_len[0] = (byte)(data_len & 0xff);
                b_data_len[1] = (byte)((data_len >> 8) & 0xff);
                b_data_len[2] = (byte)((data_len >> 16) & 0xff);
                b_data_len[3] = (byte)((data_len >> 24) & 0xff);

                if (b_data_len.Length != mSocket.Send(b_data_len))
                {
                    return false;
                }
                if(data.Length != mSocket.Send(data))
                {
                    return false;
                }
                return true;
            }
            catch (Exception)
            {
                return false;
            }
        }

        private byte[] UnpackReceiveData(byte[] data)
        {
            SHA256 sha256 = new SHA256Managed();
            int hash_size = sha256.HashSize / 8;
            //解密数据
            byte[] dec_data = mRC4.DoEnc(data);
            if (dec_data.Length > hash_size)
            {
                byte[] receive_data = new byte[dec_data.Length - hash_size];
                Array.Copy(dec_data, hash_size, receive_data, 0, receive_data.Length);
                //校验hash
                byte[] hash = sha256.ComputeHash(receive_data);
                sha256.Clear();
                if (__MemoryCompare(hash, dec_data, hash.Length))
                {
                    //解压缩
                    return DecompressData(receive_data);
                }
            }
            return null;
        }

        private byte[] PackSendData(byte[] data)
        {
            //压缩
            byte[] zip_data = CompressData(data);

            //计算hash
            SHA256 sha256 = new SHA256Managed();
            byte[] hash = sha256.ComputeHash(zip_data);
            sha256.Clear();

            byte[] send_data = new byte[hash.Length + zip_data.Length];
            hash.CopyTo(send_data, 0);
            zip_data.CopyTo(send_data, hash.Length);

            //加密数据
            byte[] enc_data = mRC4.DoEnc(send_data);
            return enc_data;
        }

        private byte[] CompressData(byte[] inData)
        {
            using (MemoryStream outMemoryStream = new MemoryStream())
            using (ZOutputStream outZStream = new ZOutputStream(outMemoryStream, zlibConst.Z_DEFAULT_COMPRESSION))
            using (Stream inMemoryStream = new MemoryStream(inData))
            {
                CopyStream(inMemoryStream, outZStream);
                outZStream.finish();
                return outMemoryStream.ToArray();
            }
        }

        private byte[] DecompressData(byte[] inData)
        {
            using (MemoryStream outMemoryStream = new MemoryStream())
            using (ZOutputStream outZStream = new ZOutputStream(outMemoryStream))
            using (Stream inMemoryStream = new MemoryStream(inData))
            {
                CopyStream(inMemoryStream, outZStream);
                outZStream.finish();
                return outMemoryStream.ToArray();
            }
        }

        private void CopyStream(System.IO.Stream input, System.IO.Stream output)
        {
            byte[] buffer = new byte[2048];
            int len;
            while ((len = input.Read(buffer, 0, buffer.Length)) > 0)
            {
                output.Write(buffer, 0, len);
            }
            output.Flush();
        }
    }
}
